src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -272,14 +322,103 @@ def upgrade_start(self, image: str, version: str) -> str:
         self.upgrade_state = UpgradeState(
             target_name=target_name,
-            progress_id=str(uuid.uuid4())
+            progress_id=str(uuid.uuid4()),
+            daemon_types=daemon_types,
+            hosts=hosts,
+            services=services,
+            total_count=limit,
+            remaining_count=limit,
         )
         self._update_upgrade_progress(0.0)
         self._save_upgrade_state()
         self._clear_upgrade_health_checks()
         self.mgr.event.set()
         return 'Initiating upgrade to %s' % (target_name)
 
+    def _validate_upgrade_filters(self, target_name: str, daemon_types: Optional[List[str]] = None, hosts: Optional[List[str]] = None, services: Optional[List[str]] = None) -> None:
+        def _latest_type(dtypes: List[str]) -> str:
+            # [::-1] gives the list in reverse
+            for daemon_type in CEPH_UPGRADE_ORDER[::-1]:
+                if daemon_type in dtypes:
+                    return daemon_type
+            return 'N/A'
+
+        def _get_earlier_daemons(dtypes: List[str], candidates: List[DaemonDescription]) -> List[DaemonDescription]:
+            latest = _latest_type(dtypes)
+            if latest == 'N/A':
+                return []
+            earlier_types = '|'.join(CEPH_UPGRADE_ORDER).split(latest)[0].split('|')[:-1]
+            earlier_types = [t for t in earlier_types if t not in dtypes]
+            return [d for d in candidates if d.daemon_type in earlier_types]
+
+        if self.upgrade_state:
+            raise OrchestratorError(
+                'Cannot set values for --daemon-types, --services or --hosts when upgrade already in progress.')
+        try:
+            target_id, target_version, target_digests = self.mgr.wait_async(
+                CephadmServe(self.mgr)._get_container_image_info(target_name))
+        except OrchestratorError as e:
+            raise OrchestratorError(f'Failed to pull {target_name}: {str(e)}')
+        # what we need to do here is build a list of daemons that must already be upgraded
+        # in order for the user's selection of daemons to upgrade to be valid. for example,
+        # if they say --daemon-types 'osd,mds' but mons have not been upgraded, we block.
+        daemons = [d for d in self.mgr.cache.get_daemons(
+        ) if d.daemon_type not in MONITORING_STACK_TYPES]
+        err_msg_base = 'Cannot start upgrade. '
+        # "dtypes" will later be filled in with the types of daemons that will be upgraded with the given parameters
+        dtypes = []
+        if daemon_types is not None:
+            dtypes = daemon_types
+            if hosts is not None:
+                dtypes = [_latest_type(dtypes)]
+                other_host_daemons = [
+                    d for d in daemons if d.hostname is not None and d.hostname not in hosts]
+                daemons = _get_earlier_daemons(dtypes, other_host_daemons)
+            else:
+                daemons = _get_earlier_daemons(dtypes, daemons)
```

Comment:
This code portion seems to be mostly the same in the three different cases. Maybe it could be factorized out from the if clauses. Se we only keep the `dtypes` calculation which seems to be different for each case.

Reply:
I tried looking through the 3 cases to see which subsections could maybe be their own function. The only thing I could ifnd really repeated was
```
                other_host_daemons = [
                    d for d in daemons if d.hostname is not None and d.hostname not in hosts]
                daemons = _get_earlier_daemons(dtypes, other_host_daemons)
            else:
                daemons = _get_earlier_daemons(dtypes, daemons)
```
but even then it's a bit annoying to try to split that section off into its own function as dtypes is being set differently in each case (within the if block in one case, outside in  the other) and that behavior would have to be preserved. What did you have in mind here in terms of splitting things up?

Reply:
I was not referring to get each section out in a separate function but re-arranging the code to make it clear that what really changes from each if clause is the `dtypes` calculation. I worked out a little big the logic and got the following, Ithink it's equivalent. BTW: I don't understand why in `elif services is not None` we don't use the `_latest_type(dtypes)` meanwhile in the other cases we do use it. Just a comment to make sure we are not forgetting to apply it for that case.

```

        if daemon_types is not None:
            dtypes = daemon_types if hosts is not None else [_latest_type(dtypes)]
            err_msg_base += 'Daemons with types earlier in upgrade order than given types need upgrading.\n'
        elif services is not None:
            # for our purposes here we can effectively convert our list of services into the
            # set of daemon types the services contain. This works because we don't allow --services
            # and --daemon-types at the same time and we only allow services of the same type
            sspecs = [self.mgr.spec_store[s].spec for s in services if self.mgr.spec_store[s].spec is not None]
            stypes = list(set([s.service_type for s in sspecs]))
            if len(stypes) != 1:
                raise OrchestratorError('Doing upgrade by service only support services of one type at a time. Found service types: {stypes}')
            for stype in stypes:
                dtypes += orchestrator.service_to_daemon_types(stype)
            dtypes = list(set(dtypes))
            err_msg_base += 'Daemons with types earlier in upgrade order than daemons from given services need upgrading.\n'
        elif hosts is not None:
            # hosts must be handled a bit differently. For this, we really need to find all the daemon types
            # that reside on hosts in the list of hosts we will upgrade. Then take the type from
            # that list that is latest in the upgrade order and check if any daemons on hosts not in the
            # provided list of hosts have a daemon with a type earlier in the upgrade order that is not upgraded.
            dtypes = list(set([d.daemon_type for d in daemons if d.daemon_type is not None and d.hostname in hosts]))
            dtypes = _latest_type(dtypes)
            err_msg_base += 'Daemons with types earlier in upgrade order than daemons on given host need upgrading.\n'

        other_host_daemons = daemons
        if hosts is not None:
            other_host_daemons = [d for d in daemons if d.hostname is not None and d.hostname not in hosts]
        daemons = _get_earlier_daemons(dtypes, other_host_daemons)

```

Reply:
it's unnecessary to do that for services because we only allow services of one daemon type at a time with that flag. The reason being allowing services of different types makes the verification significantly more complex.

Reply:
got it, thanks for the clarification :+1: